src/core/thread/key_manager.hpp

```diff
@@ -139,9 +140,20 @@ class SecurityPolicy : public otSecurityPolicy, public Equatable<SecurityPolicy>
  *
  */
 OT_TOOL_PACKED_BEGIN
-class NetworkKey : public otNetworkKey, public Equatable<NetworkKey>
+class NetworkKey : public otNetworkKey, public Equatable<NetworkKey>, public Clearable<NetworkKey>
 {
 public:
+    /**
+     * Type of Crypto used by the platform. This defines if the key is stored as a literal string, or as
+     * a reference.
+     */
+    Mac::CryptoType mCryptoType;
+
+    /**
+     * Reference to Network Key.
+     */
+    otNetworkKeyRef mKeyRef;
```

Comment:
Here, we are defining `NetworkKey` as a sub-class of `otNetworkKey` and also add additional member variables `mCryptoType` and `mKeyRef` to the `NetworkKey`.

In the code (e.g., in `thread_api.cpp` in `otThreadSetNetworkKey()`) there are situation where we cast from an `otNetworkKey` instance (base type instance) and treat it as `NetworkKey` which would not be correct then. I highly advice against this model (even if we won't be using the extra member variable on sub-class after casting, it is not a good idea to do so and it easy to make a mistake here).

-----

The general model we use in OT:
- The public OT API and type definition are all C based (say `otIp6Address` defined as `struct` and are all PoD (or standard layout).
- Then inside OT core, we define a C++ class equivalent (say `Ip6::Address` which adds methods on the type, but never add new member variable.
- This then allows us to safely cast between the C definitions and C++ core equivalents.

I highly suggest we continue to follow same model, i.e. `NetworkKey` should be mirror `otNetworkKey` and contain no additional data fields.

 If we need additional member variable we can considering defining a new class/type which is not a sub-class of `otNetworKey` but can store either a `NetworKkey or a `KeyRef` and we provide methosd to set it from a given `NetworKey` or `KeyRef`. For example:

```c++
typedef otNetworkKeyRef NetworkKeyRef;

class NetworkKeyInfo
{
public:
    void SetFrom(NetworkKey &aKey);
    void SetFrom(NetworkKeyRef aKeyRef);

    Mac::CryptoType  mCryptoType;
    NetworkKey       mLiteralKey;
    NetworkKeyRef    mKeyRef;
};
```

Reply:
Ok. I will change these.

Reply:
@abtink Doesn't that mean key_manager now has 2 copies of `NetworkKey`? I would much rather prefer to not have a new class, instead  just add  member for `NetworkKeyRef`.

Reply:
@abtink is there any specific reason to include `SetFrom` inside `NetworkKeyInfo`? I was thinking of leaving the Setters as they are today (`SetNetworkKey` to set the literal string and `SetNetworkKeyRef` for references, which currently sit out of `NetworkKey` class).

Reply:
It's just a suggestion. It's up to you to decide.
 
My idea was not to replace the `SetNetworkKey()` or `SetNetworkKeyRef()` but just adding helper setter method in the new type itself so that we ensure when we set it we also update `mCryptoType` accordingly. It can be done different ways as well.